### ISSUE TITLE
Does Vigipy implement the GPS or MGPS?

### ISSUE THREAD

**Issue by griffin-farrow** [Shakesbeery/vigiPy#8]
Hi there! Thanks for providing this phenomenal repo and I appreciate it's a long shot asking this because I suspect this repo is no longer being watched, but I had a question about whether it is the GPS or MGPS algorithm that has been implemented in Vigipy. 

The documentation says MGPS, but I noticed that the input data is only single device-event/drug-event pairs, and I'm wondering whether the code would still work if you had multi-item drug-drug-event pairs or something similar. 

Cheers for any help in advcance!

**Comment by Shakesbeery** [comment]
Hello! Lucky you I have nothing better to do than watch old repos. 😆 

Unfortunately I am in the middle of a vacation and do not have a ton of time, so I will attempt to give a quick response that will be as helpful as possible. If it fails to satisfy, then I should be able to give a more comprehensive answer in about a week when I return.

MGPS is simply the GPS algorithm extended to allow for multi-item comparisons. This repository has implemented the extended version of GPS, so although the function name remains `gps`, the internal algorithm supports MGPS. Nevertheless, you are correct in identifying that the current input structure really only supports a single column input, even when there are potential multiple interactions. This necessitates a change in data pre-processing to support the MGPS in its full capacity. Let's say that you had a dataset with possible interaction terms:

| **Drug 1 ID** | **Drug 2 ID** | **Adverse Event ID** | **Observed Count** | **Total Reports for Drug 1** | **Total Reports for Drug 2** | **Total Reports for Event** | **Total Reports** |
|---------------|---------------|----------------------|--------------------|------------------------------|------------------------------|-----------------------------|-------------------|
| Drug A        | Drug B        | Event X              | 3                  | 50                           | 60                           | 100                         | 1000              |
| Drug A        | Drug C        | Event Y              | 2                  | 50                           | 40                           | 30                          | 1000              |
| Drug B        | Drug D        | Event X              | 1                  | 60                           | 20                           | 100                         | 1000              |


In order to properly process this data while considering the interactions of different drugs, we can combine the multiple drugs into a single column. The end results would look something like:

| **Drug Combination** | **Adverse Event ID** | **Observed Count** | **Total Reports for Drug Combination** | **Total Reports for Event** | **Total Reports** |
|----------------------|----------------------|--------------------|----------------------------------------|-----------------------------|-------------------|
| Drug A + Drug B      | Event X              | 3                  | 50                                     | 100                         | 1000              |
| Drug A + Drug B      | Event Y              | 5                  | 50                                     | 30                          | 1000              |
| Drug A + Drug C      | Event X              | 0                  | 40                                     | 100                         | 1000              |
| Drug A + Drug C      | Event Z              | 2                  | 40                                     | 20                          | 1000              |
| Drug B + Drug D      | Event X              | 1                  | 30                                     | 100                         | 1000              |

The analysis has lost nothing by doing so and as long as you account for any changes in combination counts, the results should be identical to any other multi-column MGPS library. What I should _probably_ do is add a utility that does this kind of combinatorial data preparation for you.

I hope this helps! Reach out with any other questions you may have and I will endeavor to answer them as time allows.

tl;dr: Yes, you can absolutely do MGPS with this repo, but you'll have to prepare your data appropriately.

**Comment by griffin-farrow** [comment]
This is actually really useful, thanks for your help! I'm in the process of doing that data combinatorial preparation, so if I get anywhere with that, I might lodge it as a pull request!

One other question I had - I noticed that when I used Vigipy on the default dataset (in `./test/fixtures/sample.csv`) with all default values, it predicts nonsense answers for the hyperparameters (notably predicts the mixture fraction should be 33000!), I'm assuming this is just a bad initial guess for the hyperparameters problem, but I was wondering if you'd seen that before?

**Comment by Shakesbeery** [comment]
Great! I'm glad this was of some use. 

I just whipped out a quick idea for handling combinations:
```python
import pandas as pd
from itertools import chain, combinations

def generate_combinations(df, target_cols=["drug"]):
    # Initialize an empty list to store the result
    result = []
    
    # Iterate over each row in the dataframe
    for index, row in df.iterrows():
        # Extract product namesfrom the current row
        names = [row[x] for x in target_cols]
        # Get all unique combinations of names (without repetition)
        combos = list(chain.from_iterable(combinations(names, r) for r in range(1, len(names)+1)))
        # Append combinations to the result list with the other column info
        for combo in combos:
            new_data = {idx: row[idx] for idx in row.index if idx not in target_cols}
            new_data["product_combo"] = f"{'-'.join(combo)}"
            result.append(new_data)

    # Convert the result list to a new dataframe
    new_df = pd.DataFrame(result)
    
    return new_df

# Example usage:
data = {
    'drug1': ['Drug X', 'Drug Y', 'Drug Z'],
    'drug2': ['Drug A', 'Drug A', 'Drug B'],
    'drug3': ['Drug C', 'Drug Z', 'Drug R'],
    'AE': ["Death", "Vomiting", "Nausea"],
    'date': ['2024-01-01', '2024-01-02', '2024-01-03']
}
df = pd.DataFrame(data)
new_df = generate_combinations(df, target_cols=["drug1", "drug2","drug3"])
print(new_df)
```

It's still missing the logic needed for counts, but if you have more free time than me in the near future you can maybe look at adding that logic?

Another thing you may want to check out: I just pushed a new experimental branch to the repo. You can `git checkout lasso` after fetching and it supports using LASSO for disproportionality analyses. Theoretically this approach is a good fit to your use case because it uses binary matrices to track products and adverse events. This means the combinatorial nature of your data is baked into the data preparation. Your first shot might be something like this:

```python
from vigipy import lasso, convert_binary

data = pd.read_csv("YOUR_DATA.csv")
new = convert_binary(data, product_label=["drug1", "drug2", "drug3"])
results = lasso(new, alpha=.01, use_IC=True, IC_criterion="aic")
print(results.all_signals)
```

Note that this is experimental and I've done only the bare minimum testing. Also, the expectations, approach and assumptions are completely different from all the other DA functions, so the returned data is also different in structure. If you want to toy around a little and give any feedback, now is a great time to do so.

Lastly, could you provide the code you're running that's causing issues with the test sample? The sample data was never meant to be actual DA data, so it violates a lot of statistical assumptions - this might be an explanation as to why results seem janky. But there also might actually be numerical instability somewhere in the code - either way, worth checking out if you can provide the code to reproduce the issue.

**Comment by griffin-farrow** [comment]
Hey! I will take a look at the Lasso method, that seems really useful.
The code I'm using that seems to be running into a numerical instability is this:
```
import pandas as pd 
from vigipy import * 

df = pd.read_csv('test/fixtures/sample.csv')
vigipy_data = convert(df)

# calculate the GPS score
results = gps(vigipy_data, min_events=5, decision_metric='rank', decision_thres=1)

print(results.param)
# for my case, this gives {'prior_param': array([   45451.75359432,   224500.26237345, -8097268.47449176, -356162.36615722,    93911.35787145])}
```

If I've changed the optimization method in the scipy.optimize.minimize call to Nelder-Mead, you do get a totally different final value that looks a bit more sensible; but hard to know if this is a dataset issue like you say!

**Comment by Shakesbeery** [comment]
You're correct! The default minimization algorithm is occasionally failing to converge or running into numerical instability issues. 
 The `lasso` branch has now been updated to address this by defaulting to the conjugte gradient method and also exposing a new argument `minimization_method`. Any of the scipy minimization methods can be passed here. Thanks to your stability report, I also discovered that the hyperparameters were not being applied correctly after the update. I would recommend defaulting to the `lasso` branch for your development purposes until I have an opportunity to merge into master.

**Comment by griffin-farrow** [comment]
Okay, great - I'll use the lasso branch from now on. I notice that in the code, you set the hyperparameters to eps if they are less than 0, and to 1 if they are greater than 1. I'm wondering if it wouldn't be safer to instead use the `scipy` bounds argument in the minimisation call to just restrict the parameter space for the optimisation?

**Comment by Shakesbeery** [comment]
That was definitely a consideration I had and kind of waffled on a bit before implementing it in this ugly fashion. Currently `α` and `β` are only constrained to be greater than the machine epsilon to prevent log errors. `w` is constrained to:` 0 ≤ w ≤ 1`.  I chose this to avoid unnecessarily restricting the available solvers the user now has access to via the `minimization_method` argument since `bounds` only works for a subset of the minimization methods. 

However... As I wrote the above paragraph I realized how trivial it would be to handle that. I think it's a good idea to expose the bounds argument as well and add a check for compatibility with the selected solver method.

**Comment by Shakesbeery** [comment]
I also just pushed a major change that includes a multi-item preprocessor. I whipped it out pretty fast, so there's no guarantee I did any of it right. Seems to work fine though, so let me know if it reproduces the results you are looking for.

**Comment by Shakesbeery** [comment]
@griffin-farrow - This was a fantastic issue to create and it has led to the largest single improvement of the repo since it was created, so thank you! We now have the `convert_multi_item()` function which technically addresses the original purpose of this thread. I'm going to work on getting things updated for merge into the master branch and then close this thread. If you discover any shortcomings of the multi-item approach as developed, post it here, but I'm planning on closing this issue on Wednesday/Thursday. Naturally, any other questions/ideas/problems you have at any point from here can always be used to create a new issue.

**Comment by griffin-farrow** [comment]
@Shakesbeery I'm very glad! I will be using this update and I will let you know if I come across anything with it!

**Comment by Shakesbeery** [comment]
lasso branch now merged into master and you should be good to use v1.4 from here on out.
